salt/transport/ipc.py

```diff
@@ -582,11 +583,116 @@ def __del__(self):
+        try:
+            yield self._read_in_progress.acquire(timeout=0)
+        except tornado.gen.TimeoutError:
+            raise tornado.gen.Return(None)
+
+        log.debug('IPC Subscriber Service is starting reading')
+        # If timeout is not specified we need to set some here to make the service able to check
+        # is there any handler waiting for data.
+        if timeout is None:
+            timeout = 5
+
+        read_stream_future = None
+        while self._has_subscribers():
+            if read_stream_future is None:
+                read_stream_future = self.stream.read_bytes(4096, partial=True)
+
+            try:
+                wire_bytes = yield FutureWithTimeout(self.io_loop,
+                                                     read_stream_future,
+                                                     timeout)
+                read_stream_future = None
+
+                self.unpacker.feed(wire_bytes)
+                msgs = [msg['body'] for msg in self.unpacker]
+                self._feed_subscribers(msgs)
+            except TornadoTimeoutError:
+                # Continue checking are there alive waiting handlers
+                # Keep 'read_stream_future' alive to wait it more in the next loop
+                continue
+            except tornado.iostream.StreamClosedError as exc:
+                log.trace('Subscriber disconnected from IPC %s', self.socket_path)
+                self._feed_subscribers([None])
+                break
+            except Exception as exc:
+                log.error('Exception occurred in Subscriber while handling stream: %s', exc)
+                self._feed_subscribers([exc])
+                break
+
+        log.debug('IPC Subscriber Service is stopping due to a lack of subscribers')
+        self._read_in_progress.release()
+        raise tornado.gen.Return(None)
+
+    @tornado.gen.coroutine
+    def read(self, handler, timeout=None):
+        '''
+        Asynchronously read messages and invoke a callback when they are ready.
+
+        :param callback: A callback with the received data
+        '''
+        self._subscribe(handler)
+        while not self.connected():
+            try:
+                yield self.connect(timeout=5)
+            except tornado.iostream.StreamClosedError:
+                log.trace('Subscriber closed stream on IPC %s before connect', self.socket_path)
+                yield tornado.gen.sleep(1)
+            except Exception as exc:
+                log.error('Exception occurred while Subscriber connecting: %s', exc)
+                yield tornado.gen.sleep(1)
+        self._read(timeout)
+
+    def close(self):
+        '''
+        Routines to handle any cleanup before the instance shuts down.
+        Sockets and filehandles should be closed explicitly, to prevent
+        leaks.
+        '''
+        if not self._closing:
+            super(IPCMessageSubscriberService, self).close()
+
+    def __del__(self):
```

Comment:
@DmitryKuzmenko Why do we need to use the `__del__` method? Is it possible to explicitly close these where we use them? Perhaps in a process teardown?

Reply:
Currently it's common way to close all the transport channels, so my opinion if we'd like to change this it shall be done everywhere in a specific PR. I've just leaved things here as they were before.